File: app/src/main/java/com/waveneuro/ui/user/password/reset/ResetPasswordActivity.java
package com.waveneuro.ui.user.password.reset;

import android.app.AlertDialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.text.Editable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.ForegroundColorSpan;
import android.text.style.UnderlineSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Button;
import android.widget.TextView;

import androidx.lifecycle.Observer;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputLayout;
import com.mobsandgeeks.saripaar.annotation.NotEmpty;
import com.waveneuro.R;
import com.waveneuro.ui.base.BaseFormActivity;
import com.waveneuro.ui.dashboard.more.WebCommand;
import com.waveneuro.ui.user.email.forgot.ForgotUsernameCommand;
import com.waveneuro.ui.user.login.LoginCommand;
import com.waveneuro.ui.user.password.code.ForgotPasswordCodeCommand;
import com.waveneuro.ui.user.password.password.first.SetPasswordViewCommand;
import com.waveneuro.ui.user.password.recovery.RecoveryInstructionsCommand;
import com.waveneuro.ui.user.registration.RegistrationCommand;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ResetPasswordActivity extends BaseFormActivity {

    @NotEmpty(trim = true, message = "Enter email")
    @BindView(R.id.et_username)
    TextInputLayout etUsername;

    @BindView(R.id.tv_log_in)
    TextView tvLogIn;

    @BindView(R.id.btn_send_reset_link)
    MaterialButton btnSendLink;

    @Inject
    ResetPasswordViewModel resetPasswordViewModel;

    @Inject
    ForgotUsernameCommand forgotUsernameCommand;
    @Inject
    RegistrationCommand registrationCommand;
    @Inject
    RecoveryInstructionsCommand recoveryInstructionsCommand;
    @Inject
    ForgotPasswordCodeCommand forgotPasswordCodeCommand;

    @Inject
    LoginCommand loginCommand;
    @Inject
    SetPasswordViewCommand setPasswordViewCommand;

    @Inject
    WebCommand webCommand;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        getWindow().requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS);
        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
        getWindow().setStatusBarColor(Color.TRANSPARENT);

        super.onCreate(savedInstanceState);
        activityComponent().inject(this);
        setContentView(R.layout.activity_reset_password);
        ButterKnife.bind(this);

        setView();
        setObserver();
    }

    private void setView() {
        logInSpanText();
        setupInputWatcher();
    }

    private void logInSpanText() {
        SpannableString spannableString = new SpannableString(getString(R.string.log_in));

        spannableString.setSpan(new UnderlineSpan(), 0, 6, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

        spannableString.setSpan(new ForegroundColorSpan(getResources().getColor(R.color.yellow_dim)), 0, 6, Spannable.SPAN_EXCLUSIVE_INCLUSIVE);
        tvLogIn.setText(spannableString);
    }

    private void setupInputWatcher() {
        etUsername.getEditText().addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                btnSendLink.setEnabled(count != 0);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void setObserver() {
        this.resetPasswordViewModel.getData().observe(this, resetPasswordViewStateObserver);
        this.resetPasswordViewModel.getViewEffect().observe(this, resetPasswordViewEffectObserver);
    }

    Observer<ResetPasswordViewState> resetPasswordViewStateObserver = viewState -> {
        if (viewState instanceof ResetPasswordViewState.Success) {
            ResetPasswordViewState.Success success = (ResetPasswordViewState.Success) viewState;
            launchCheckEmailDialog();
        } else if (viewState instanceof ResetPasswordViewState.Failure) {
            ResetPasswordViewState.Failure error = (ResetPasswordViewState.Failure) viewState;
            onFailure(error.getError());
        } else if (viewState instanceof ResetPasswordViewState.Loading) {
            ResetPasswordViewState.Loading loading = ((ResetPasswordViewState.Loading) viewState);
            if (loading.getLoading())
                displayWait("Loading...", null);
            else
                removeWait();
        }
    };

    Observer<ResetPasswordViewEffect> resetPasswordViewEffectObserver = viewEffect -> {
        if (viewEffect instanceof ResetPasswordViewEffect.BackRedirect) {
            goBack();
        } else if (viewEffect instanceof ResetPasswordViewEffect.LoginRedirect) {
            launchLoginScreen();
        }
    };

    private void launchLoginScreen() {
        loginCommand.navigate();
        finish();
    }

    private void launchCheckEmailDialog() {
        final AlertDialog.Builder builder = new AlertDialog.Builder(this, R.style.PopUp);
        ViewGroup viewGroup = findViewById(android.R.id.content);
        View dialogView = LayoutInflater.from(this).inflate(R.layout.dialog_popup, viewGroup, false);
        TextView tvTitle = dialogView.findViewById(R.id.tv_title);
        TextView tvContent = dialogView.findViewById(R.id.tv_content);
        Button btnPrimary = dialogView.findViewById(R.id.btn_primary);
        tvTitle.setText(R.string.check_your_email);
        tvTitle.setTextSize(24);
        tvContent.setText(R.string.recovery_info);
        btnPrimary.setText(R.string.open_email_app);
        builder.setView(dialogView);
        final AlertDialog ad = builder.create();
        btnPrimary.setOnClickListener(v -> {
            ad.dismiss();
            forgotPasswordCodeCommand.navigate(etUsername.getEditText().getText().toString().trim());
            try {
                Intent intent = new Intent(Intent.ACTION_MAIN);
                intent.addCategory(Intent.CATEGORY_APP_EMAIL);
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                this.startActivity(intent);
            } catch (android.content.ActivityNotFoundException e) {
            }
        });
        ad.show();
    }

    @OnClick(R.id.tv_log_in)
    public void onClickBack() {
        loginCommand.navigate();
    }

    private void goBack() {
        onBackPressed();
    }

    @Override
    public void submit() {
        this.resetPasswordViewModel.processEvent(
                new ResetPasswordViewEvent.ForgotPasswordClicked(
                        etUsername.getEditText().getText().toString().trim()));
    }

    @OnClick(R.id.btn_send_reset_link)
    public void onClickSendResetLink() {
        this.mValidator.validate();
    }

}

File: app/src/main/java/com/waveneuro/ui/user/password/reset/ResetPasswordActivity.kt
package com.waveneuro.ui.user.password.reset

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.*
import android.text.style.ForegroundColorSpan
import android.text.style.UnderlineSpan
import android.view.View
import android.view.Window
import androidx.activity.viewModels
import androidx.core.view.isVisible
import androidx.lifecycle.Observer
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.waveneuro.R
import com.waveneuro.databinding.ActivityResetPasswordBinding
import com.waveneuro.databinding.DialogPopupBinding
import com.waveneuro.ui.base.activity.BaseViewModelActivity
import com.waveneuro.ui.user.login.LoginActivity
import com.waveneuro.ui.user.login.LoginViewEffect
import com.waveneuro.ui.user.password.code.ForgotPasswordCodeActivity
import com.waveneuro.ui.user.password.reset.ResetPasswordViewEffect.LoginRedirect
import com.waveneuro.ui.user.password.reset.viewmodel.ResetPasswordViewModel
import com.waveneuro.ui.user.password.reset.viewmodel.ResetPasswordViewModelImpl
import com.waveneuro.utils.ext.getAppComponent
import com.waveneuro.utils.ext.toast

class ResetPasswordActivity : BaseViewModelActivity<ActivityResetPasswordBinding, ResetPasswordViewModel>() {

    override val viewModel: ResetPasswordViewModelImpl by viewModels {
        getAppComponent().resetPasswordViewModelFactory()
    }

    override fun initBinding(): ActivityResetPasswordBinding =
        ActivityResetPasswordBinding.inflate(layoutInflater)

    override fun onCreate(savedInstanceState: Bundle?) {
        window.requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS)
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN)
        window.statusBarColor = Color.TRANSPARENT
        super.onCreate(savedInstanceState)

        setView()
        setObserver()
    }

    private fun setView() {
        with(binding) {
            tvLogIn.setOnClickListener {
                startActivity(LoginActivity.newIntent(this@ResetPasswordActivity))
            }
            btnSendResetLink.setOnClickListener {
                if (tipUsername.text.isNullOrBlank()) {
                    toast("Enter email")
                } else {
                    viewModel.processEvent(ResetPasswordViewEvent.ForgotPasswordClicked(
                        tipUsername.text.toString().trim()
                    ))
                }
            }
        }
        logInSpanText()
        setupInputWatcher()
    }

    private fun logInSpanText() {
        val spannableString = SpannableString(getString(R.string.log_in))
        spannableString.setSpan(UnderlineSpan(), 0, 6, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        spannableString.setSpan(
            ForegroundColorSpan(resources.getColor(R.color.yellow_dim)),
            0,
            6,
            Spannable.SPAN_EXCLUSIVE_INCLUSIVE
        )
        binding.tvLogIn.text = spannableString
    }

    private fun setupInputWatcher() {
        binding.etUsername.editText?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                binding.btnSendResetLink.isEnabled = count != 0
            }

            override fun afterTextChanged(s: Editable) {}
        })
    }

    private fun setObserver() {
        viewModel.viewEffect.observe(this, Observer { viewEffect ->
            when (viewEffect) {
                is ResetPasswordViewEffect.BackRedirect -> {
                    goBack()
                }
                is LoginRedirect -> {
                    launchLoginScreen()
                }
                is ResetPasswordViewEffect.Success -> {
                    launchCheckEmailDialog()
                }
                is ResetPasswordViewEffect.ShowErrorDialog -> {
                    showErrorDialog(viewEffect.title ?: "", viewEffect.message)
                }
            }
        })
    }

    private fun launchLoginScreen() {
        startActivity(LoginActivity.newIntent(this@ResetPasswordActivity))
        finish()
    }

    private fun launchCheckEmailDialog() {
        val dialogBinding = DialogPopupBinding.inflate(layoutInflater)
        val builder = MaterialAlertDialogBuilder(this, R.style.PopUp).setView(dialogBinding.root)
        val dialog = builder.create()

        with(dialogBinding) {
            ibtnClose.isVisible = true
            tvTitle.setText(R.string.check_your_email)
            tvContent.setText(R.string.recovery_info)
            btnPrimary.setText(R.string.open_email_app)
            btnPrimary.setOnClickListener {
                dialog.dismiss()
                startActivity(ForgotPasswordCodeActivity.newIntent(
                    this@ResetPasswordActivity,
                    binding.tipUsername.text.toString().trim()
                ))

                try {
                    val intent = Intent(Intent.ACTION_MAIN)
                    intent.addCategory(Intent.CATEGORY_APP_EMAIL)
                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    this@ResetPasswordActivity.startActivity(intent)
                } catch (e: ActivityNotFoundException) { }
            }
            ibtnClose.setOnClickListener {
                dialog.dismiss()
                startActivity(ForgotPasswordCodeActivity.newIntent(
                    this@ResetPasswordActivity,
                    binding.tipUsername.text.toString().trim()
                ))
            }
        }

        dialog.show()
    }

    private fun goBack() {
        onBackPressed()
    }

    companion object {
        fun newIntent(context: Context) = Intent(context, ResetPasswordActivity::class.java)
    }

}